File: Stage-Leoni/Backend/controllers/DepartmentController.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Contrôleur pour les départements
"""

from flask import jsonify, request
from models.Department import Department

class DepartmentController:
    def __init__(self):
        self.department_model = Department()
    
    def get_all_departments(self):
        """Récupérer tous les départements"""
        try:
            print("🔍 CONTROLLER_DEPARTMENTS: Récupération de tous les départements", flush=True)
            
            result = self.department_model.find_all(active_only=True)
            
            if result['success']:
                print(f"✅ CONTROLLER_DEPARTMENTS: {result['count']} départements trouvés", flush=True)
                return jsonify({
                    'success': True,
                    'departments': result['departments'],
                    'count': result['count']
                }), 200
            else:
                print(f"❌ CONTROLLER_DEPARTMENTS: {result['message']}", flush=True)
                return jsonify({
                    'success': False,
                    'message': result['message'],
                    'departments': [],
                    'count': 0
                }), 500
                
        except Exception as e:
            print(f"❌ CONTROLLER_DEPARTMENTS: Exception = {str(e)}", flush=True)
            return jsonify({
                'success': False,
                'message': 'Erreur lors de la récupération des départements',
                'error': str(e),
                'departments': [],
                'count': 0
            }), 500
        finally:
            self.department_model.close_connection()
    
    def get_departments_by_location(self, location_id):
        """Récupérer les départements par location"""
        try:
            print(f"🔍 CONTROLLER_DEPARTMENTS: Récupération des départements pour location {location_id}", flush=True)
            
            result = self.department_model.find_by_location(location_id)
            
            if result['success']:
                print(f"✅ CONTROLLER_DEPARTMENTS: {result['count']} départements trouvés pour la location", flush=True)
                return jsonify({
                    'success': True,
                    'departments': result['departments'],
                    'count': result['count'],
                    'location_id': location_id
                }), 200
            else:
                print(f"❌ CONTROLLER_DEPARTMENTS: {result['message']}", flush=True)
                return jsonify({
                    'success': False,
                    'message': result['message'],
                    'departments': [],
                    'count': 0,
                    'location_id': location_id
                }), 500
                
        except Exception as e:
            print(f"❌ CONTROLLER_DEPARTMENTS: Exception = {str(e)}", flush=True)
            return jsonify({
                'success': False,
                'message': 'Erreur lors de la récupération des départements',
                'error': str(e),
                'departments': [],
                'count': 0,
                'location_id': location_id
            }), 500
        finally:
            self.department_model.close_connection()
    
    def get_department_by_id(self, department_id):
        """Récupérer un département par ID"""
        try:
            print(f"🔍 CONTROLLER_DEPARTMENTS: Récupération du département {department_id}", flush=True)
            
            result = self.department_model.find_by_id(department_id)
            
            if result['success']:
                print(f"✅ CONTROLLER_DEPARTMENTS: Département trouvé", flush=True)
                return jsonify({
                    'success': True,
                    'department': result['department']
                }), 200
            else:
                print(f"❌ CONTROLLER_DEPARTMENTS: {result['message']}", flush=True)
                return jsonify({
                    'success': False,
                    'message': result['message']
                }), 404
                
        except Exception as e:
            print(f"❌ CONTROLLER_DEPARTMENTS: Exception = {str(e)}", flush=True)
            return jsonify({
                'success': False,
                'message': 'Erreur lors de la récupération du département',
                'error': str(e)
            }), 500
        finally:
            self.department_model.close_connection()
    
    def create_department(self):
        """Créer un nouveau département"""
        try:
            data = request.get_json()
            
            if not data or 'name' not in data:
                return jsonify({
                    'success': False,
                    'message': 'Le nom du département est requis'
                }), 400
            
            print(f"🔍 CONTROLLER_DEPARTMENTS: Création du département {data['name']}", flush=True)
            
            result = self.department_model.create(
                name=data['name'],
                description=data.get('description'),
                location_ref=data.get('locationRef'),
                active=data.get('active', True)
            )
            
            if result['success']:
                print(f"✅ CONTROLLER_DEPARTMENTS: Département créé avec ID {result['department_id']}", flush=True)
                return jsonify({
                    'success': True,
                    'message': result['message'],
                    'department_id': result['department_id']
                }), 201
            else:
                print(f"❌ CONTROLLER_DEPARTMENTS: {result['message']}", flush=True)
                return jsonify({
                    'success': False,
                    'message': result['message']
                }), 500
                
        except Exception as e:
            print(f"❌ CONTROLLER_DEPARTMENTS: Exception = {str(e)}", flush=True)
            return jsonify({
                'success': False,
                'message': 'Erreur lors de la création du département',
                'error': str(e)
            }), 500
        finally:
            self.department_model.close_connection()
    
    def update_department(self, department_id):
        """Mettre à jour un département"""
        try:
            data = request.get_json()
            
            if not data:
                return jsonify({
                    'success': False,
                    'message': 'Aucune donnée fournie'
                }), 400
            
            print(f"🔍 CONTROLLER_DEPARTMENTS: Mise à jour du département {department_id}", flush=True)
            
            result = self.department_model.update(
                department_id=department_id,
                name=data.get('name'),
                description=data.get('description'),
                location_ref=data.get('locationRef'),
                active=data.get('active')
            )
            
            if result['success']:
                print(f"✅ CONTROLLER_DEPARTMENTS: Département mis à jour", flush=True)
                return jsonify({
                    'success': True,
                    'message': result['message']
                }), 200
            else:
                print(f"❌ CONTROLLER_DEPARTMENTS: {result['message']}", flush=True)
                return jsonify({
                    'success': False,
                    'message': result['message']
                }), 404 if 'non trouvé' in result['message'] else 500
                
        except Exception as e:
            print(f"❌ CONTROLLER_DEPARTMENTS: Exception = {str(e)}", flush=True)
            return jsonify({
                'success': False,
                'message': 'Erreur lors de la mise à jour du département',
                'error': str(e)
            }), 500
        finally:
            self.department_model.close_connection()
    
    def delete_department(self, department_id):
        """Supprimer un département"""
        try:
            print(f"🔍 CONTROLLER_DEPARTMENTS: Suppression du département {department_id}", flush=True)
            
            result = self.department_model.delete(department_id)
            
            if result['success']:
                print(f"✅ CONTROLLER_DEPARTMENTS: Département supprimé", flush=True)
                return jsonify({
                    'success': True,
                    'message': result['message']
                }), 200
            else:
                print(f"❌ CONTROLLER_DEPARTMENTS: {result['message']}", flush=True)
                return jsonify({
                    'success': False,
                    'message': result['message']
                }), 404 if 'non trouvé' in result['message'] else 500
                
        except Exception as e:
            print(f"❌ CONTROLLER_DEPARTMENTS: Exception = {str(e)}", flush=True)
            return jsonify({
                'success': False,
                'message': 'Erreur lors de la suppression du département',
                'error': str(e)
            }), 500
        finally:
            self.department_model.close_connection()

File: Stage-Leoni/LeoniApp/src/controllers/DepartmentController.js
import LocationService from '../services/LocationService';
import { DepartmentModel } from '../models/DepartmentModel';

/**
 * Contrôleur pour gérer les départements
 */
export class DepartmentController {
  constructor() {
    // LocationService utilise des méthodes statiques, pas besoin d'instanciation
  }

  /**
   * Récupérer tous les départements
   * @returns {Promise<DepartmentModel[]>}
   */
  async getAllDepartments() {
    try {
      console.log('🏢 DepartmentController: Récupération de tous les départements');
      
      const response = await LocationService.getAllDepartments();
      
      if (response.success && response.departments) {
        const departments = DepartmentModel.fromAPIArray(response.departments);
        console.log(`✅ DepartmentController: ${departments.length} départements récupérés`);
        return departments;
      } else {
        console.warn('⚠️ DepartmentController: Réponse API invalide', response);
        return [];
      }
    } catch (error) {
      console.error('❌ DepartmentController: Erreur lors de la récupération des départements:', error);
      throw new Error(`Erreur lors du chargement des départements: ${error.message}`);
    }
  }

  /**
   * Récupérer les départements d'une location spécifique
   * @param {string} locationId 
   * @returns {Promise<DepartmentModel[]>}
   */
  async getDepartmentsByLocation(locationId) {
    try {
      if (!locationId) {
        console.warn('⚠️ DepartmentController: ID de location manquant');
        return [];
      }

      console.log(`🏢 DepartmentController: Récupération des départements pour la location ${locationId}`);
      
      const response = await LocationService.getDepartmentsByLocation(locationId);
      
      if (response.success && response.departments) {
        console.log(`🔍 DepartmentController: Données brutes reçues:`, response.departments);
        const departments = DepartmentModel.fromAPIArray(response.departments);
        console.log(`✅ DepartmentController: ${departments.length} départements récupérés pour la location`);
        console.log(`🔍 DepartmentController: Premier département transformé:`, departments[0]);
        return departments;
      } else {
        console.warn('⚠️ DepartmentController: Aucun département trouvé pour cette location', locationId);
        return [];
      }
    } catch (error) {
      console.error('❌ DepartmentController: Erreur lors de la récupération des départements:', error);
      throw new Error(`Erreur lors du chargement des départements: ${error.message}`);
    }
  }

  /**
   * Récupérer un département par son ID
   * @param {string} departmentId 
   * @returns {Promise<DepartmentModel|null>}
   */
  async getDepartmentById(departmentId) {
    try {
      if (!departmentId) {
        console.warn('⚠️ DepartmentController: ID de département manquant');
        return null;
      }

      console.log(`🏢 DepartmentController: Récupération du département ${departmentId}`);
      
      // Pour l'instant, on récupère tous les départements et on filtre
      // On pourrait améliorer cela avec un endpoint spécifique
      const allDepartments = await this.getAllDepartments();
      const department = allDepartments.find(dept => dept.id === departmentId);
      
      if (department) {
        console.log(`✅ DepartmentController: Département ${department.name} récupéré`);
        return department;
      } else {
        console.warn('⚠️ DepartmentController: Département non trouvé', departmentId);
        return null;
      }
    } catch (error) {
      console.error('❌ DepartmentController: Erreur lors de la récupération du département:', error);
      throw new Error(`Erreur lors du chargement du département: ${error.message}`);
    }
  }

  /**
   * Récupérer les départements actifs seulement
   * @param {string|null} locationId - Optionnel: filtrer par location
   * @returns {Promise<DepartmentModel[]>}
   */
  async getActiveDepartments(locationId = null) {
    try {
      let departments;
      
      if (locationId) {
        departments = await this.getDepartmentsByLocation(locationId);
      } else {
        departments = await this.getAllDepartments();
      }
      
      const activeDepartments = DepartmentModel.filterActive(departments);
      
      console.log(`✅ DepartmentController: ${activeDepartments.length} départements actifs sur ${departments.length}`);
      return activeDepartments;
    } catch (error) {
      console.error('❌ DepartmentController: Erreur lors de la récupération des départements actifs:', error);
      throw error;
    }
  }

  /**
   * Formater les départements pour un picker/dropdown
   * @param {string|null} locationId - Optionnel: filtrer par location
   * @returns {Promise<Array>}
   */
  async getDepartmentsForPicker(locationId = null) {
    try {
      const departments = await this.getActiveDepartments(locationId);
      const sortedDepartments = DepartmentModel.sortByName(departments);
      
      const pickerData = sortedDepartments.map(department => ({
        _id: department.id,
        name: department.getDisplayName(),
        code: department.code,
        locationId: department.locationRef || department.locationId,
        locationName: department.getLocationName()
      }));

      console.log(`✅ DepartmentController: ${pickerData.length} départements formatés pour picker`);
      return pickerData;
    } catch (error) {
      console.error('❌ DepartmentController: Erreur lors du formatage pour picker:', error);
      return [];
    }
  }

  /**
   * Rechercher des départements par nom
   * @param {string} searchTerm 
   * @param {string|null} locationId - Optionnel: filtrer par location
   * @returns {Promise<DepartmentModel[]>}
   */
  async searchDepartments(searchTerm, locationId = null) {
    try {
      if (!searchTerm || searchTerm.trim().length === 0) {
        return locationId ? 
          await this.getDepartmentsByLocation(locationId) : 
          await this.getAllDepartments();
      }

      const departments = locationId ? 
        await this.getDepartmentsByLocation(locationId) : 
        await this.getAllDepartments();
        
      const filteredDepartments = departments.filter(department =>
        department.name.toLowerCase().includes(searchTerm.toLowerCase()) ||
        department.code.toLowerCase().includes(searchTerm.toLowerCase()) ||
        (department.description && department.description.toLowerCase().includes(searchTerm.toLowerCase()))
      );

      console.log(`🔍 DepartmentController: ${filteredDepartments.length} départements trouvés pour "${searchTerm}"`);
      return filteredDepartments;
    } catch (error) {
      console.error('❌ DepartmentController: Erreur lors de la recherche:', error);
      return [];
    }
  }

  /**
   * Valider les données d'un département
   * @param {Object} departmentData 
   * @returns {Object} {isValid, errors}
   */
  validateDepartmentData(departmentData) {
    const errors = [];

    if (!departmentData.name || departmentData.name.trim().length === 0) {
      errors.push('Le nom du département est obligatoire');
    }

    if (!departmentData.locationRef && !departmentData.locationId) {
      errors.push('Le site du département est obligatoire');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  /**
   * Grouper les départements par location
   * @param {DepartmentModel[]} departments 
   * @returns {Object} Départements groupés par location
   */
  groupDepartmentsByLocation(departments) {
    if (!Array.isArray(departments)) return {};

    const grouped = {};
    
    departments.forEach(department => {
      const locationKey = department.locationRef || department.locationId || 'unknown';
      const locationName = department.getLocationName();
      
      if (!grouped[locationKey]) {
        grouped[locationKey] = {
          locationId: locationKey,
          locationName: locationName,
          departments: []
        };
      }
      
      grouped[locationKey].departments.push(department);
    });

    // Trier les départements dans chaque groupe
    Object.values(grouped).forEach(group => {
      group.departments = DepartmentModel.sortByName(group.departments);
    });

    console.log(`📊 DepartmentController: Départements groupés en ${Object.keys(grouped).length} locations`);
    return grouped;
  }

  /**
   * Nettoyer le cache des départements
   */
  clearCache() {
    console.log('🧹 DepartmentController: Cache des départements nettoyé');
    // Ici on pourrait implémenter un système de cache si nécessaire
  }
}

export default DepartmentController;
